src/mbgl/renderer/layers/render_raster_layer.cpp

```diff
@@ -255,40 +250,206 @@ void RenderRasterLayer::update(gfx::ShaderRegistry& shaders,
+
+                vertices.emplace_back(std::array<int16_t, 2>{{tlVertex.x, brVertex.y}});
+                attributes.emplace_back(std::array<int16_t, 2>{{tlVertex.x, brVertex.y}});
+
+                vertices.emplace_back(std::array<int16_t, 2>{{brVertex.x, brVertex.y}});
+                attributes.emplace_back(std::array<int16_t, 2>{{brVertex.x, brVertex.y}});
+
+                auto& segment = segments.back();
+                assert(segment.vertexLength <= std::numeric_limits<uint16_t>::max());
+                const auto offset = static_cast<uint16_t>(segment.vertexLength);
+
+                // 0, 1, 2
+                // 1, 2, 3
+                indices.insert(indices.end(),
+                               {offset, static_cast<uint16_t>(offset + 1), static_cast<uint16_t>(offset + 2u)});
+                indices.insert(indices.end(),
+                               {static_cast<uint16_t>(offset + 1),
+                                static_cast<uint16_t>(offset + 2),
+                                static_cast<uint16_t>(offset + 3)});
+
+                segment.vertexLength += vertexLength;
+                segment.indexLength += 6;
+            }
+        };
+
+        std::vector<std::array<int16_t, 2>> vertices, attributes;
+        std::vector<uint16_t> indices;
+        std::vector<Segment<void>> segments;
+        buildRenderData(bucket.mask, vertices, attributes, indices, segments);
+        builder->addVertices(vertices, 0, vertices.size());
+        builder->setSegments(gfx::Triangles(), indices, segments);
+
+        // attributes
+        {
+            gfx::VertexAttributeArray vertexAttrs;
+            if (auto& attr = vertexAttrs.getOrAdd("a_texture_pos")) {
+                std::size_t index{0};
+                for (auto& a : attributes) {
+                    attr->set<gfx::VertexAttribute::int2>(index++, {a[0], a[1]});
+                }
+            }
+            builder->setVertexAttributes(std::move(vertexAttrs));
         }
-    } else if (renderTiles) {
-        tileLayerGroup->observeDrawables([&](gfx::UniqueDrawable& drawable) {
-            // Has this tile dropped out of the cover set?
-            if (const auto it = std::find_if(renderTiles->begin(),
-                                             renderTiles->end(),
-                                             [&drawable](const auto& renderTile) {
-                                                 return drawable->getTileID() == renderTile.get().getOverscaledTileID();
-                                             });
-                it == renderTiles->end()) {
-                // remove it
-                drawable.reset();
-                ++stats.tileDrawablesRemoved;
+
+        // textures
+        std::shared_ptr<gfx::Texture2D> tex0 = context.createTexture2D();
+        tex0->setImage(bucket.image);
+        auto location0 = rasterShader->getSamplerLocation("u_image0");
+        if (location0.has_value()) {
```

Comment:
We can check this first and skip allocating a texture instance if the shader is missing a sampler for whatever reason.